### PR TITLE
Replace Knox with aws-sdk

### PR DESCRIPTION
**Pull request by mikehazell**
### Overview

The Knox library which this package was previously based on has gone unmaintained for some time and is now failing in many scenarios. This version replaces Knox with the official [AWS Javascript SDK](https://aws.amazon.com/sdk-for-node-js/).

### Breaking changes

The option `headers` has been replaced with `uploadParams`. If you were setting custom headers with previous version of the S3 Storage Adapter you will need to change these to use the appropriate `param` as defined in the [AWS Documentation](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload)

For example, `{ headers: { 'x-amz-acl': 'public-read' } }` should now be `{ uploadParams: { ACL: 'public-read' } }`.

### Additions

- **publicUrl**: You can now customise the public url by passing either a domain name as a string (eg. `{ publicUrl: "https://xxxxxx.cloudfront.net" }`) or by passing a function which takes the `file` object and returns a the url as a string.
```js
{ publicUrl: file => `https://xxxxxx.cloudfront.net${file.path}/${file.filename}` }
```

### Other

- **path**: The requirement for `path` to have a **leading slash** has been removed. The previous implementation failed to catch this miss-configuration and Knox helpfully made the file uploads work anyway. This has lead to a situation where it is possible/likely that there are existing installations where a miss-configured path is stored in the database. To avoid breaking these installs we now handle adding or removing the leading slash as required.

### Thanks
Thanks to @aaronfranco, @stevenkaspar, @alsoicode and @sktt who all had a go at this refactor before me. I have borrowed heavily from all of your work.

### CODE REVIEW

**Comment by alsoicode** [comment]
+1 for this. Nice work.

File: index.js [hunk 1]
```diff
@@ -69,28 +101,29 @@ S3Adapter.SCHEMA_FIELD_DEFAULTS = {
...
-S3Adapter.prototype._knoxForFile = function (file) {
-	// Clients are allowed to store the bucket name in the file structure. If they
-	// do it'll make it possible to have some files in one bucket and some files
-	// in another bucket. The knox client is configured per-bucket, so if you're
-	// using multiple buckets we'll need a different knox client for each file.
-	if (file.bucket && file.bucket !== this.options.bucket) {
-		var s3options = assign({}, this.options, { bucket: file.bucket });
-		return knox.createClient(s3options);
+S3Adapter.prototype._resolveBucket = function (file) {
+	if (file && file.bucket && file.bucket) {
```
Comment (dominikwilkowski) [inline index.js]: probably don't need `file.bucket` twice here?

File: package.json [hunk 2]
```diff
@@ -10,8 +10,8 @@
   },
```
Comment (dominikwilkowski) [inline package.json:14]: We should version bump the package no? `"version": "2.0.0",` on line 3.

**Review by dominikwilkowski — CHANGES_REQUESTED** [review]
Looks awesome, Thanks @mikehazell; love the comments!

**Review by dominikwilkowski — APPROVED** [review]

**Pull request merged**